### DISCUSSION TITLE
Probable Nutils patch interface bug

### DISCUSSION THREAD

**Discussion by pabigu** [evalf/nutils#857] [Q&A] [answered]
Hello,

We are trying to create a multipatch geometry. However, we are facing an unexpected interface orientation error.
The error appears for mesh.multipatch() :
Exception has occurred: NotImplementedError
patch interfaces must have the same order of axes and the same orientation per axis

Despite multiple orientation checks within the problematic interface, the error persists. We think it is probably a bug in Nutils, but you may have an other idea about the origin of this problem.
Here is the code to reproduce the error :

```
from nutils import mesh, export

patches = [[2,4,3,7,0,8,1,11],
           [5,9,6,10,4,8,7,11]
           ]

domain, param = mesh.multipatch(patches=patches,
                                patchverts=[[0,0,1],[1,0,1],[0,1/3,2/3],[2/3,1/3,2/3],[0,2/3,2/3],[0,2/3,1/3],
                                            [1/3,2/3,1/3],[2/3,2/3,2/3],[0,1,1],[0,1,0],[1/4,1,1/4],[1,1,1]
                                            ], nelems={None: 1}
                                )

bezier = domain.sample('bezier', 2)
points = bezier.eval(param)
export.vtk("your file export directory", bezier.tri, points) # add the directory to which your file will be exported
```
Here is a view of the mutlipatch definition :
![multipatch definition](https://github.com/evalf/nutils/assets/157494025/b81765d8-69ec-43ca-b01a-482a96900a4f)

Best,
Dimitri

**Comment by gertjanvanzwieten** [comment, marked as answer]
The multipatch construction is currently needlessly restrictive, in that it requires the (in this case) 2D local coordinate systems of the matching faces to coincide, and rejects faces for which the first and second dimension are swapped. This is not at all a fundamental issue, merely a practical one, but it will take some effort to resolve, and we will most likely spend this effort in the planned rewrite of the multipatch framework. In the meantime I hope the additional restrictions are not prohibitive to creating the desired multi-patch topology. For the provided minimal example, which has a matching face in [4,7,8,11], the fix is as simple as changing the configuration to `patches = [[2,4,3,7,0,8,1,11], [5,6,9,10,4,7,8,11]]`, i.e. flipping the first 2 dimensions of the second patch - the faces maintain the order of the dimensions they inherit from the parent.

**Reply by gertjanvanzwieten** [reply]
PS sorry for the late reply!

**Reply by pabigu** [reply]
Thanks to your help, the error no longer raised in `mesh.multipatch` but we have the exact same error raised in `domain.interfaces['interpatch']`

Here is the code to reproduce the error :
```
from nutils import mesh

patchSet = [[10, 12, 6, 8, 11, 1, 7, 9],
            [10, 12, 6, 8, 0, 1, 2, 3],
            [13, 15, 10, 12, 14, 5, 11, 1],
            [13, 15, 10, 12, 4, 5, 0, 1]]
coord = ((-49.99999999999997, 46.46446609406726, -203.5355339059327),
         (-49.99999999999997, -1.4210854715202e-14, -250.0),
         (-49.99999999999997, 53.53553390593274, -203.5355339059327),
         (-49.99999999999997, 100.0, -250.0),
         (-49.99999999999997, 46.46446609406726, -196.4644660940673),
         (-49.99999999999997, -7.105427357601002e-15, -150.0),
         (-12.1446609406726, 37.85533905932743, -203.5355339059328),
         (3.535533905932788, 1.407731819508732e-14, -203.5355339059328),
         (50.0, 100.0, -250.0),
         (50.00000000000001, 1.224646799147354e-14, -250.0),
         (-17.14466094067261, 32.85533905932742, -203.5355339059328),
         (-3.53553390593271, 1.778729211085775e-14, -203.5355339059328),
         (-27.50000000000001, 22.50000000000001, -250.0),
         (-17.14466094067261, 32.85533905932742, -196.4644660940673),
         (-3.53553390593271, 1.778729211085775e-14, -196.4644660940673),
         (-27.50000000000001, 22.50000000000001, -150.0))

domain, param = mesh.multipatch(patches=patchSet,patchverts=coord,nelems={None: 1})

domain.interfaces['interpatch'] # >>NotImplementedError: patch interfaces must have the same order of axes and the same orientation per axis
```
To understand where the error came from, we printed the variables involving in the raising statement in `__inti__()` from the `MultipatchTopology` class in `nutils\topology.py` like this :
```
    def __init__(self, patches: Sequence[Patch]):
        assert isinstance(patches, Sequence) and all(isinstance(patch, Patch) for patch in patches), f'patches={patches!r}'
        self.patches = tuple(patches)

        space = patches[0].topo.space
        assert all(patch.topo.space == space for patch in patches)

        for boundaryid, patchdata in self._patchinterfaces.items():
            if len(patchdata) == 1:
                continue
            transposes = set()
            reverses = set()
            for topo, boundary in patchdata:
                assert boundary.transpose[-1] == boundary.dim
                transposes.add(tuple(i-1 if i > boundary.dim else i for i in boundary.transpose[:-1]))
                reverses.add(boundary.reverse[:boundary.dim]+boundary.reverse[boundary.dim+1:])
            print('boundary :',boundary) # added print
            print('transposes :',transposes) # added print
            print('reverses :',reverses) # added print
            print('') # added print
            if len(transposes) != 1 or len(reverses) != 1:
                print('patchdata :',patchdata) # added print
                raise NotImplementedError('patch interfaces must have the same order of axes and the same orientation per axis')

        super().__init__(
            space,
            util.sum(patch.topo.references for patch in self.patches),
            transformseq.chain([patch.topo.transforms for patch in self.patches], self.patches[0].topo.transforms.todims, self.patches[0].topo.ndims),
            transformseq.chain([patch.topo.opposites for patch in self.patches], self.patches[0].topo.transforms.todims, self.patches[0].topo.ndims))
```
Just before the error occurs, this returns :
```
boundary : PatchBoundary(id=(10, 12), dim=1, side=0, reverse=(False, False), transpose=(0, 1))
transposes : {(0,)}
reverses : {(False,), (True,)}

patchdata : ((<nutils.topology.TransformChainsTopology object at 0x0000014D9BCC7AA0>, PatchBoundary(id=(10, 12), dim=1, side=-1, reverse=(False, True), transpose=(0, 1))), (<nutils.topology.TransformChainsTopology object at 0x0000014D9BD0C4A0>, PatchBoundary(id=(10, 12), dim=1, side=-1, reverse=(True, True), transpose=(0, 1))), (<nutils.topology.TransformChainsTopology object at 0x0000014D9BD0D220>, PatchBoundary(id=(10, 12), dim=1, side=-1, reverse=(False, True), transpose=(0, 1))), (<nutils.topology.TransformChainsTopology object at 0x0000014D9BD0D3D0>, PatchBoundary(id=(10, 12), dim=1, side=0, reverse=(False, False), transpose=(0, 1))))
```
So the problem seems to come from the edge (10,12) shared by 4 patches. But we haven't been able to understand the meaning of the `side`, `reverse` and `transpose` variables. This prevent us from finding out what is wrong with the patch definition.

**Reply by gertjanvanzwieten** [reply]
It seems that there was indeed a bug in the construction of interfaces, thanks for bringing that to light. Could you see if the problem goes away in the [multipatch](https://github.com/evalf/nutils/tree/multipatch) branch, or [bpo](https://github.com/evalf/nutils/tree/bpo) for a backport to stable? It should, as it includes your problem as unit test `test_topology.multipatch_misc.test_partial_ring`!

**Reply by pabigu** [reply]
Hi, good news, the problem raised with the master branch disappears in both multipatch and bpo branches.

**Reply by gertjanvanzwieten** [reply]
Great, thanks for checking! We just merged the changes in the master branch and will soon merge the backport in version 8 (keep an eye on #873 to know when that happens). Let us know if there are other things you run into.

**Comment by pabigu** [comment]
Hi, thank you for your reply.
I agree, I've just tested with local coordinates systems correspondence at interfaces in addition to only the orientations correspondence on various examples and no errors were raised.
